platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/models/ProjectStateStore.kt

```diff
@@ -93,33 +100,80 @@ abstract class ProjectStateStore<K, V>(
         }
     }
 
+    /**
+     * Create or load value, with load-after-store semantics
+     */
     fun loadOrCreateValue(key: K, creator: () -> V): V {
-        val addressOfCached = locateCachedValue(key)
-        if (addressOfCached != null) {
-            try {
-                return valuesStore.read(addressOfCached)
-            } catch (e: Exception) {
-                throw RuntimeException("Could not load entry for $key", e)
-            }
+        val address = loadOrCreateAddress(key, creator)
+        return readValue(key, address)
+    }
+
+    /**
+     * If value has to be created, the original value is returned without (de)serialization
+     *
+     * This is delicate API, because the original value is returned only for the first call of this function.
+     * As such, the call should probably be guarded with additional synchronization and memoization.
+     *
+     * Prefer [loadOrCreateValue] unless there is a special reason.
+     */
+    fun loadOrCreateOriginalValue(key: K, creator: () -> V): V {
+        val originalValueCapture = AtomicReference<V>()
```

Comment:
💅 It seems a regular `var` would be simpler since there's no sharing involved?

```kotlin
var originalValueCapture: V? = null
```

Reply:
Taking the current calls to this method into account, I would agree.

At the same time, locally, the capturing lambda could still be called from a different thread if it happens to call `finalizeIfNotAlready` before the current thread. And even though there are probably enough happens-before relations to guarantee it would work, I still prefer to keep the concurrent nature of this capturing more explicit.

On a related note, there is something "shady" about Kotlin wrapping local variables in `Ref<>` and friends to make them mutable in lambdas, and I don't like it 😅 It's over that line for me, where Kotlin is too implicit

Reply:
> the capturing lambda could still be called from a different thread if it happens to call finalizeIfNotAlready before the current thread. 

Right 👍